pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,954 @@
+		return err
+	}
+
+	// Update the current effective BGPPolicy state.
+	curState.routes = curRoutes
+	curState.peerConfigs = curPeerConfigs
+	c.bgpPolicyState = curState
+
+	return nil
+}
+
+func getPeerConfigsForKeys(peerKeys sets.Set[string], allPeerConfigs map[string]bgp.PeerConfig) []bgp.PeerConfig {
+	peerConfigs := make([]bgp.PeerConfig, 0, len(peerKeys))
+	for peer := range peerKeys {
+		peerConfigs = append(peerConfigs, allPeerConfigs[peer])
+	}
+	return peerConfigs
+}
+
+func (c *Controller) reconcileBGPPeers(curPeerConfigs, prePeerConfigs map[string]bgp.PeerConfig, bgpServer bgp.Interface, bgpServerUpdated bool) error {
+	prePeerKeys := sets.KeySet(prePeerConfigs)
+	curPeerKeys := sets.KeySet(curPeerConfigs)
+
+	var peerToAddKeys sets.Set[string]
+	if !bgpServerUpdated {
+		peerToAddKeys = curPeerKeys.Difference(prePeerKeys)
+	} else {
+		peerToAddKeys = curPeerKeys
+	}
+	peerConfigsToAdd := getPeerConfigsForKeys(peerToAddKeys, curPeerConfigs)
+	for _, peer := range peerConfigsToAdd {
+		if err := bgpServer.AddPeer(c.ctx, peer); err != nil {
+			return err
+		}
+	}
+
+	if !bgpServerUpdated {
+		peerToUpdateKeys := sets.New[string]()
+		remainPeerKeys := prePeerKeys.Intersection(curPeerKeys)
+		for peerKey := range remainPeerKeys {
+			prevPeerConfig := prePeerConfigs[peerKey]
+			curPeerConfig := curPeerConfigs[peerKey]
+			if !reflect.DeepEqual(prevPeerConfig, curPeerConfig) {
+				peerToUpdateKeys.Insert(peerKey)
+			}
+		}
+		peerToUpdateConfigs := getPeerConfigsForKeys(peerToUpdateKeys, curPeerConfigs)
+		for _, peer := range peerToUpdateConfigs {
+			if err := bgpServer.UpdatePeer(c.ctx, peer); err != nil {
+				return err
+			}
+		}
+
+		peerToDeleteKeys := prePeerKeys.Difference(curPeerKeys)
+		peerToDeleteConfigs := getPeerConfigsForKeys(peerToDeleteKeys, prePeerConfigs)
+		for _, peer := range peerToDeleteConfigs {
+			if err := bgpServer.RemovePeer(c.ctx, peer); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Controller) reconcileRoutes(curRoutes, preRoutes sets.Set[bgp.Route], bgpServer bgp.Interface, bgpServerUpdated bool) error {
+	var routesToAdvertise sets.Set[bgp.Route]
+	if !bgpServerUpdated {
+		routesToAdvertise = curRoutes.Difference(preRoutes)
+	} else {
+		routesToAdvertise = curRoutes
+	}
+	if routesToAdvertise.Len() != 0 {
+		if err := bgpServer.AdvertiseRoutes(c.ctx, routesToAdvertise.UnsortedList()); err != nil {
+			return err
+		}
+	}
+
+	if !bgpServerUpdated {
+		routesToWithdraw := preRoutes.Difference(curRoutes)
+		if routesToWithdraw.Len() != 0 {
+			if err := bgpServer.WithdrawRoutes(c.ctx, routesToWithdraw.UnsortedList()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) getRouterID() (string, error) {
+	var routerID string
+	// For IPv6 only environment, the BGP routerID should be specified by K8s Node annotation `antrea.io/bgp-route-id`.
```

Comment:
could you elaborate this more? I don't know why it needs a manually specified router id only for IPv6.

s/bgp-route-id/bgp-router-id/

Reply:
Will add.

Reply:
Added some comments:

```
	// According to RFC 4271:
	// BGP Identifier:
	//   This 4-octet unsigned integer indicates the BGP Identifier of
	//   the sender.  A given BGP speaker sets the value of its BGP
	//   Identifier to an IP address that is assigned to that BGP
	//   speaker.  The value of the BGP Identifier is determined upon
	//   startup and is the same for every local interface and BGP peer.
	// In goBGP, only an IPv4 address can be used as the BGP Identifier (BGP router ID). For IPv4-only or dual-stack
	// Kubernetes clusters, we use the Node's IPv4 address as the BGP router ID, ensuring uniqueness. For IPv6-only
	// Kubernetes clusters, where no Node IPv4 address is available, the BGP router ID must be specified via the Node
	// annotation `node.antrea.io/bgppolicy-router-id`.
```